From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 25 Dec 2022 23:39:05 +0100
Subject: [PATCH] Ignore null legacy structure data

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Add null check in LegacyStructureDataHandler"
By: PureGero <puregero@gmail.com>
As part of: MultiPaper (https://github.com/MultiPaper/MultiPaper)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

diff --git a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
index 1611aa47d47d38b9aec65de98064b7982fc514c8..40dc72a415037683c06ff79975a0daee1d93e084 100644
--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -208,13 +208,20 @@ public class LegacyStructureDataHandler implements LegacyTagFixer {
 
     private synchronized void populateCaches(DimensionDataStorage storage) {
         if (!this.cachesInitialized) {
+            boolean ignoreNullLegacyStructureData = org.galemc.gale.configuration.GaleGlobalConfiguration.get().misc.ignoreNullLegacyStructureData; // Gale - MultiPaper - ignore null legacy structure data
             for (String string : this.legacyKeys) {
                 CompoundTag compoundTag = new CompoundTag();
 
                 try {
-                    compoundTag = storage.readTagFromDisk(string, DataFixTypes.SAVED_DATA_STRUCTURE_FEATURE_INDICES, 1493)
+                    // Gale start - MultiPaper - ignore null legacy structure data
+                    CompoundTag tag = storage.readTagFromDisk(string, DataFixTypes.SAVED_DATA_STRUCTURE_FEATURE_INDICES, 1493);
+
+                    if (ignoreNullLegacyStructureData && tag == null) continue;
+
+                    compoundTag = tag
                         .getCompoundOrEmpty("data")
                         .getCompoundOrEmpty("Features");
+                    // Gale end - MultiPaper - ignore null legacy structure data
                     if (compoundTag.isEmpty()) {
                         continue;
                     }
